Support for advanced generics

alexeev509 opened the pull request:
Was added supporting of generic types (declaring and usage):

1. Declaring generic functions and procs.
2. Usage of the generic procs.
3. Declaring generic actions.
4. Usage generic actions.
5. Declaring generic typealias.
6. Creating generic automatons
7. Usage generic pros's of automatons.
8. Creating objects of generic automatons.
9. Usage is/as with generic types.
11. Declaring generic types.
12. Added keywords: in, out, where, ?.

Was added class: class GenericType.
Were added generic test's (see generic.lsl).
Was added field TypeReference in the TypeReference class.

alexeev509 committed:
Was created plain supported of generics for fun's;

alexeev509 committed:
In out supporting was added;

alexeev509 committed:
Where supporting was added;

alexeev509 committed:
Refactoring of parametrized fun's for "generic" keyword

alexeev509 committed:
Supporting of the generics for types;

alexeev509 committed:
Were added type bounds for types;

alexeev509 committed:
Was added where section for types;

alexeev509 committed:
Was added generic types for automatons

alexeev509 committed:
Were added generics for fun's; Plain realization;

alexeev509 committed:
Were added generics for proc's; Plain realization;

alexeev509 committed:
Were added generics for actions;

alexeev509 committed:
QualifiedAccessUtils little fix;

alexeev509 committed:
Little refactoring;

1. Was deleted class Generic
2. It was replaced by GenericType for TypeDefBlock

alexeev509 committed:
Was added supporting of the procs usage with generics;

alexeev509 committed:
Little fixes for test;

alexeev509 committed:
Supporting of the generic automatons constructor invokes inside fun’s and inside procs;

alexeev509 committed:
Some improving of the generics;

alexeev509 committed:
Action usage was added;

alexeev509 committed:
Target parametrized test;

alexeev509 committed:
Is test was added;

alexeev509 committed:
As test was added;

alexeev509 committed:
In out of typeDefBlock was refactored;

alexeev509 committed:
Undounded wildcard of call constructor;

alexeev509 committed:
Undounded wildcard tests were added;

alexeev509 committed:
Were added test-cases for generics usage with bounds;

alexeev509 committed:
Were added typealias generic supporting;

alexeev509 committed:
Little fix for dump of the generics;

alexeev509 committed:
Was changed structure of generics in AntLr parser;

alexeev509 committed:
Was fixed where section;

alexeev509 committed:
Little refactoring of the same parts;

alexeev509 committed:
Little refactoring, mainly dumpers;

alexeev509 committed:
Return types dump fixes;

alexeev509 committed:
Little refactoring;

alexeev509 committed:
Little refactoring;

alexeev509 committed:
AVLTree from ArkTs stdlib partly was described (for generics testing)

dog-m commented on src/test/testdata/expected/lsl/AVLTreeArkTsFunSignatures.lsl:
```diff
@@ -0,0 +1,56 @@
+libsl "1.1.0";
+library std version "11" language "Java" url "https://gitee.com/openharmony/arkcompiler_runtime_core/blob/master/static_core/plugins/ets/stdlib/std/containers/AVLTree.ets";
```
I strongly suggest removing any mention of third parties, especially trademarks and other IPs (intellectual properties).

```suggestion
library std version "11" language "unknown" url "-";
```

alexeev509 replied:
Fixed

dog-m commented on src/test/testdata/expected/lsl/generics.lsl:
```diff
@@ -0,0 +1,95 @@
+libsl "1.1.0";
+library std version "11" language "Java" url "-";
```
```suggestion
library std version "11" language "unknown" url "-";
```

alexeev509 replied:
Fixed

dog-m commented on src/test/testdata/expected/lsl/generics.lsl:
```diff
@@ -0,0 +1,95 @@
+
+        }
+        var obj: HashMap<Int, Int> = x as HashMap<Int, Int>;
+        var newHashMapUnbounded: HashMap<?, ?> = new A<?, ?>(state = Initialized);
```
I think, automata constructors should accept only real (i.e. non-wildcard) generic type arguments.

alexeev509 replied:
Fixed

dog-m commented on src/test/testdata/expected/lsl/generics.lsl:
```diff
@@ -0,0 +1,95 @@
+
+        }
+        var obj2: HashMap<?, ?> = x as HashMap<?, ?>;
+        var newHashMapBounded: HashMap<in Int, out String> = new A<in Int, out String>(state = Initialized);
```
I think, automata constructors should accept only real (i.e. non-wildcard) generic type arguments.

alexeev509 replied:
Fixed

dog-m commented on src/test/testdata/lsl/AVLTreeArkTsFunSignatures.lsl:
```diff
@@ -0,0 +1,105 @@
+
+library std
+    version "11"
+    language "Java"
+    url "https://gitee.com/openharmony/arkcompiler_runtime_core/blob/master/static_core/plugins/ets/stdlib/std/containers/AVLTree.ets";
```
I strongly suggest removing any mention of third parties, especially trademarks and other IPs (intellectual properties).
```suggestion
    language "unknown"
    url "-";
```

alexeev509 replied:
Fixed

dog-m commented on src/test/testdata/lsl/generics.lsl:
```diff
@@ -0,0 +1,118 @@
+libsl "1.1.0";
+
+library std
+    version "11"
+    language "Java"
```
```suggestion
    version "11"
    language "unknown"
```

alexeev509 replied:
Fixed;

dog-m committed:
Added a test case for multiple generic constraints

alexeev509 committed:
Headers of lsl files fixes;

alexeev509 committed:
Merge remote-tracking branch 'origin/Alexeev-parser-improvements' into Alexeev-parser-improvements

alexeev509 committed:
Fixes for type with many constraints;

alexeev509 committed:
Fixes for type with many constraints;

alexeev509 committed:
Was fixed type of the variable which invokes constructor of automaton.

This is impossible now.
var newHashMapUnbounded: HashMap<?, ?> = new A<Int, Int>(state = Initialized);

alexeev509 committed:
Was fixed Recursive property accessor;

dog-m commented on src/main/kotlin/org/jetbrains/research/libsl/type/types.kt:
```diff
@@ -3,24 +3,25 @@ package org.jetbrains.research.libsl.type
     val name: String
     val isPointer: Boolean
     val context: LslContextBase
+    var typeBound: GenericTypeBound
```
Why a type should have a bound?
As I understand it, "type + bound" mean "set/range of types".

alexeev509 replied:
It was deleted from here;

dog-m commented on src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt:
```diff
@@ -486,10 +492,27 @@ class ExpressionVisitor(
+
+        }
+        
+        typeOfVariable?.genericReferences?.forEach {
+            if (!GenericTypeBound.EMPTY.equals(it.typeBound) || "?".equals(it.name))
+            // TODO: add for all exceptions in parser concrete places where it was appeared.
+                throw error("Type of variable which assigns to invokation of the constructor can't contain generics with wildcard int type")
+
+        }
```
I am a bit confused. Why do we need to check the variable type during automaton construction?
I think these use-cases should be disallowed:
```dart
var foo: Collection<?> = new AutoCollection<?>(state=...);
var bar: Collection<?> = new AutoCollection<in String>(state=...);
```
but these -- allowed:
```dart
var fizz: Collection<?> = new AutoCollection<Int>(state=...);
var buzz: Array<out Number> = new AutoArray<Byte>(state=...);
```
*(note: attention to the automaton construction expression, not the variable declaration part)*

alexeev509 replied:
Fixed;

alexeev509 committed:
Was deleted field typeBound from Type;

alexeev509 committed:
Type variable fixes;
